### PR TITLE
ADBDEV-3168-1: ORCA produces bogus plan for queries with CTE during handling distribution for Sequence children

### PR DESCRIPTION
**Pull request by RekGRpth**
With some conditions ORCA generates `Redistribute Motion` from all segments to one segment or generates a plan which can't be executed by reason of inconsistency between CTE producers and consumers.
For example, run query with two CTEs one of which is inlined by filter
```sql
CREATE TABLE d (a int, b int, c int) DISTRIBUTED BY (a);
CREATE TABLE r (a int, b int, c char(255)) DISTRIBUTED REPLICATED;
INSERT INTO d SELECT 1, generate_series(1,10), 1;
INSERT INTO r SELECT 1, 2, generate_series(1,100);
EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a);
```
that gives with asserts
```sql
FATAL:  Unexpected internal error (assert.c:44)
DETAIL:  FailedAssertion("!(pMotion->numHashSegments < motion_recv)", File: "explain.c", Line: 2534)
HINT:  Process 12638 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```
and without asserts gives bogus plan
```sql
 Gather Motion 1:1  (slice1; segments: 1)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  GroupAggregate
                     Group Key: d.b
                     ->  Sort
                           Sort Key: d.b
                           ->  Redistribute Motion 3:1  (slice2; segments: 3)
                                 Hash Key: d.b
                                 Hash Module: 3
                                 ->  Seq Scan on d
         ->  Hash Join
               Hash Cond: (r.a = d_1.a)
               ->  Seq Scan on r
               ->  Hash
                     ->  Broadcast Motion 3:1  (slice3; segments: 3)
                           ->  Hash Join
                                 Hash Cond: (d_1.b = share0_ref3.b)
                                 ->  Redistribute Motion 3:3  (slice4; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
                                 ->  Hash
                                       ->  Hash Join
                                             Hash Cond: (share0_ref3.b = share0_ref2.b)
                                             ->  Shared Scan (share slice:id 3:0)
                                             ->  Hash
                                                   ->  Shared Scan (share slice:id 3:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(28 rows)
```
In this plan `Shared Scan (share slice:id 1:0)` is CTE producer that's executed only on one segment (due to `Redistribute Motion 3:1  (slice2; segments: 3)` from all segments to one segments), but CTE consumers `Shared Scan (share slice:id 3:0)` are executed on all segments and then query hangs.
Or run query with two CTEs
```sql
CREATE TABLE d (a int, b int, c int) DISTRIBUTED BY (a);
CREATE TABLE r (a int, b int, c char(255)) DISTRIBUTED REPLICATED;
INSERT INTO d SELECT 1, generate_series(1,10), 1;
INSERT INTO r SELECT 1, 2, generate_series(1,100);
EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a) JOIN h i USING (a);
```
that gives with asserts
```sql
FATAL:  Unexpected internal error (assert.c:44)
DETAIL:  FailedAssertion("!(pMotion->numHashSegments < motion_recv)", File: "explain.c", Line: 2534)
HINT:  Process 12757 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```
and without asserts gives bogus plan
```sql
 Gather Motion 1:1  (slice1; segments: 1)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  GroupAggregate
                     Group Key: d.b
                     ->  Sort
                           Sort Key: d.b
                           ->  Redistribute Motion 3:1  (slice2; segments: 3)
                                 Hash Key: d.b
                                 Hash Module: 3
                                 ->  Seq Scan on d
         ->  Sequence
               ->  Shared Scan (share slice:id 1:1)
                     ->  Hash Join
                           Hash Cond: (d_1.b = share0_ref3.b)
                           ->  Redistribute Motion 3:1  (slice3; segments: 3)
                                 Hash Key: d_1.b
                                 Hash Module: 3
                                 ->  Seq Scan on d d_1
                           ->  Hash
                                 ->  Hash Join
                                       Hash Cond: (share0_ref3.b = share0_ref2.b)
                                       ->  Shared Scan (share slice:id 1:0)
                                       ->  Hash
                                             ->  Shared Scan (share slice:id 1:0)
               ->  Hash Join
                     Hash Cond: (r.a = share1_ref3.a)
                     ->  Seq Scan on r
                     ->  Hash
                           ->  Broadcast Motion 3:1  (slice4; segments: 3)
                                 ->  Hash Join
                                       Hash Cond: (share1_ref3.a = share1_ref2.a)
                                       ->  Shared Scan (share slice:id 4:1)
                                       ->  Hash
                                             ->  Broadcast Motion 3:3  (slice5; segments: 3)
                                                   ->  Shared Scan (share slice:id 5:1)
 Optimizer: Pivotal Optimizer (GPORCA)
(37 rows)
```
In this plan `Shared Scan (share slice:id 1:0)` and `Shared Scan (share slice:id 1:1)` are CTE producers that are executed only on one segment (due to `Redistribute Motion 3:1  (slice2; segments: 3)` and `Redistribute Motion 3:1  (slice3; segments: 3)` from all segments to one segments), but CTE consumers `Shared Scan (share slice:id 1:0)` and `Shared Scan (share slice:id 4:1)` and `Shared Scan (share slice:id 5:1)` are executed on all segments and then query hangs.
We get Redistribute from all segments to one segments because the following. When translating expression to DXL, ORCA sets one segment for input array if strict or tainted replicated distribution detected. https://github.com/greenplum-db/gpdb/blob/98e79490a26d0d9db4c9239ee1c4b33d8af65ec0/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L7332-L7346 This one segment is used as output array for below motions including redistribute and broadcast.
Solution is do not allow replicated distribution when required distribution in sequence is singleton on master or non-singleton with not allowed replicated.
In case of inlining one CTE we also need to push down non-singleton with not allowed replicated through filter. Because filter is appeared when CTE may be inlined and inlining allowed or only one consumer exists for that CTE. https://github.com/greenplum-db/gpdb/blob/98e79490a26d0d9db4c9239ee1c4b33d8af65ec0/src/backend/gporca/libgpopt/src/xforms/CXformCTEAnchor2TrivialSelect.cpp#L58-L62 In this case ORCA append logical select, https://github.com/greenplum-db/gpdb/blob/98e79490a26d0d9db4c9239ee1c4b33d8af65ec0/src/backend/gporca/libgpopt/src/xforms/CXformCTEAnchor2TrivialSelect.cpp#L104-L106 which later is translated to filter. https://github.com/greenplum-db/gpdb/blob/98e79490a26d0d9db4c9239ee1c4b33d8af65ec0/src/backend/gporca/libgpopt/src/xforms/CXformSelect2Filter.cpp#L94-L95
It is gives right plans.
For two CTEs one of which is inlined by filter
```sql
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  GroupAggregate
                     Group Key: d.b
                     ->  Sort
                           Sort Key: d.b
                           ->  Redistribute Motion 3:3  (slice2; segments: 3)
                                 Hash Key: d.b
                                 ->  Seq Scan on d
         ->  Hash Join
               Hash Cond: (r.a = d_1.a)
               ->  Result
                     ->  Seq Scan on r
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: d_1.a
                           ->  Hash Join
                                 Hash Cond: (d_1.b = share0_ref3.b)
                                 ->  Redistribute Motion 3:3  (slice4; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
                                 ->  Hash
                                       ->  Hash Join
                                             Hash Cond: (share0_ref3.b = share0_ref2.b)
                                             ->  Shared Scan (share slice:id 3:0)
                                             ->  Hash
                                                   ->  Shared Scan (share slice:id 3:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(29 rows)
```
For two CTEs
```sql
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  GroupAggregate
                     Group Key: d.b
                     ->  Sort
                           Sort Key: d.b
                           ->  Redistribute Motion 3:3  (slice2; segments: 3)
                                 Hash Key: d.b
                                 ->  Seq Scan on d
         ->  Sequence
               ->  Shared Scan (share slice:id 1:1)
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: d_1.a
                           ->  Hash Join
                                 Hash Cond: (d_1.b = share0_ref3.b)
                                 ->  Redistribute Motion 3:3  (slice4; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
                                 ->  Hash
                                       ->  Hash Join
                                             Hash Cond: (share0_ref3.b = share0_ref2.b)
                                             ->  Shared Scan (share slice:id 3:0)
                                             ->  Hash
                                                   ->  Shared Scan (share slice:id 3:0)
               ->  Hash Join
                     Hash Cond: (r.a = share1_ref3.a)
                     ->  Result
                           ->  Seq Scan on r
                     ->  Hash
                           ->  Hash Join
                                 Hash Cond: (share1_ref3.a = share1_ref2.a)
                                 ->  Shared Scan (share slice:id 1:1)
                                 ->  Hash
                                       ->  Shared Scan (share slice:id 1:1)
 Optimizer: Pivotal Optimizer (GPORCA)
(36 rows)
```
6X_STABLE also is affected by this problem.

### CODE REVIEW

**Comment by RekGRpth** [comment]
description was updated

**Comment by RekGRpth** [comment]
but it is another version of https://github.com/greenplum-db/gpdb/pull/14361

File: src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp [hunk 1]
```diff
@@ -0,0 +1,703 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dgkimura) [inline src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp:1]: I suggest adding a couple more MDP tests.

1.  Test nest loop join scenario:
```
SET optimizer_enable_hashjoin=off;

EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a);
```

2.  Test hash agg scenario:
```
SET optimizer_enable_groupagg=off;

EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT * FROM r JOIN h USING (a);
```

**Review by dgkimura — APPROVED** [review]
Thanks for diving in and iterating on this solution.  Please consider adding a couple more tests that I mentioned.  Otherwise, looks good to me.

File: src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp (hunk 1 above)
Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp:1]: > 1. Test nest loop join scenario:
> 
> 
> ```
> SET optimizer_enable_hashjoin=off;
> 
> EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
> WITH e AS (
>     SELECT DISTINCT b FROM d
> ), h AS (
>     SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
> ) SELECT * FROM r JOIN h USING (a);
> ```

does not affect by this bug?
```sql
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  GroupAggregate
                     Group Key: d.b
                     ->  Sort
                           Sort Key: d.b
                           ->  Redistribute Motion 3:3  (slice2; segments: 3)
                                 Hash Key: d.b
                                 ->  Seq Scan on d
         ->  Nested Loop
               Join Filter: (r.a = d_1.a)
               ->  Seq Scan on r
               ->  Nested Loop
                     Join Filter: (d_1.b = share0_ref2.b)
                     ->  Seq Scan on d d_1
                     ->  Materialize
                           ->  Broadcast Motion 3:3  (slice3; segments: 3)
                                 ->  Nested Loop
                                       Join Filter: (share0_ref2.b = share0_ref3.b)
                                       ->  Shared Scan (share slice:id 3:0)
                                       ->  Shared Scan (share slice:id 3:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(23 rows)
```

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp:1]: > 3\. Test hash agg scenario:
> 
> 
> ```
> SET optimizer_enable_groupagg=off;
> 
> EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
> WITH e AS (
>     SELECT DISTINCT b FROM d
> ), h AS (
>     SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
> ) SELECT * FROM r JOIN h USING (a);
> ```

does affect this bug
```sql
 Gather Motion 1:1  (slice1; segments: 1)
   ->  Sequence
         ->  Shared Scan (share slice:id 1:0)
               ->  HashAggregate
                     Group Key: d.b
                     ->  Redistribute Motion 3:1  (slice2; segments: 3)
                           Hash Key: d.b
                           Hash Module: 3
                           ->  Seq Scan on d
         ->  Hash Join
               Hash Cond: (r.a = d_1.a)
               ->  Seq Scan on r
               ->  Hash
                     ->  Broadcast Motion 3:1  (slice3; segments: 3)
                           ->  Hash Join
                                 Hash Cond: (d_1.b = share0_ref3.b)
                                 ->  Redistribute Motion 3:3  (slice4; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
                                 ->  Hash
                                       ->  Hash Join
                                             Hash Cond: (share0_ref3.b = share0_ref2.b)
                                             ->  Shared Scan (share slice:id 3:0)
                                             ->  Hash
                                                   ->  Shared Scan (share slice:id 3:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(26 rows)
```

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTE15Replicated.mdp:1]: I added mpd files for hash agg case

**Comment by gpopt** [comment]
@RekGRpth  Have you rebased on latest main branch, which has fixed some flakey tests.

**Comment by RekGRpth** [comment]
> Have you rebased on latest main branch, which has fixed the following flakey tests

done

File: src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp [hunk 2]
```diff
@@ -259,6 +259,16 @@ CPhysicalSequence::PdsRequired(CMemoryPool *mp,
 			true /* fAllowReplicated */, false /* fAllowEnforced */);
 	}
 
+	if ((CDistributionSpec::EdtSingleton == pdsRequired->Edt() &&
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalSequence.cpp:262]: Can we add a comment here similar to how we have a comment above. Also, why do we need `CDistributionSpecSingleton::PdssConvert(pdsRequired)->FOnMaster()`? (can you add this reason to the code comment)

**Review by chrishajas — COMMENTED** [review]
Code looks good, but please add some comments to the code since this is a tricky area

File: src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp [hunk 3]
```diff
@@ -124,6 +125,14 @@ CPhysicalFilter::PdsRequired(CMemoryPool *mp, CExpressionHandle &exprhdl,
 		return pdsRequired;
 	}
 
+	if (CDistributionSpec::EdtNonSingleton == pdsRequired->Edt() &&
+		!CDistributionSpecNonSingleton::PdsConvert(pdsRequired)
+			 ->FAllowReplicated())
+	{
```
[the remark this answers is not in the document]
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp:131]: done

**Review by chrishajas — APPROVED** [review]

File: src/backend/gporca/libgpopt/src/operators/CPhysicalJoin.cpp [hunk 4]
```diff
@@ -496,6 +497,40 @@ CPhysicalJoin::PrsDerive(CMemoryPool *mp, CExpressionHandle &exprhdl) const
...
+	GPOS_ASSERT(nullptr != ped);
+
+	// get distribution delivered by the physical node
+	CDistributionSpec *pds = CDrvdPropPlan::Pdpplan(exprhdl.Pdp())->Pds();
+	if (ped->FCompatible(pds))
+	{
+		if (CDistributionSpec::EdtNonSingleton == ped->PdsRequired()->Edt() &&
+			!CDistributionSpecNonSingleton::PdsConvert(ped->PdsRequired())
+				 ->FAllowReplicated() &&
+			CDistributionSpec::EdtStrictReplicated == pds->Edt())
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalJoin.cpp:521]: Does it have to be strict replicated? Is it possible to be EdtTaintedReplicated?

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalJoin.cpp]: Hm... this code is unnecessary, I removed it

File: src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp [hunk 5]
```diff
@@ -124,6 +125,16 @@ CPhysicalFilter::PdsRequired(CMemoryPool *mp, CExpressionHandle &exprhdl,
...
+		// this situation arises when we have Filter instead inlined CTE,
+		// in this case, we need to push down non-singleton with not allowed replicated through Filter
+		pdsRequired->AddRef();
+		return pdsRequired;
+	}
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp:136]: Should we move this part to function `CPhysical::PdsUnary`? PhysicalFilter may not be the only operator that can see the issue.

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp:136]: I just move this code here on request https://github.com/greenplum-db/gpdb/pull/14361#discussion_r1093880170

Comment (gpopt) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalFilter.cpp:136]: Ah, ok. :)

**Review by gpopt — APPROVED** [review]

**Comment by dgkimura** [comment]
Merged. Thanks for the contribution!

**Pull request merged**